internals/cli/cmd_push.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cli
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/canonical/go-flags"
+
+	"github.com/canonical/pebble/client"
+)
+
+const cmdPushSummary = "Transfer a file to the remote system"
+const cmdPushDescription = `
+The push command transfers a file to the remote system.
+`
+
+type cmdPush struct {
+	client *client.Client
+
+	MakeDirs    bool   `short:"p"`
+	Permissions string `short:"m"`
+	UserID      *int   `long:"uid"`
+	User        string `long:"user"`
+	GroupID     *int   `long:"gid"`
+	Group       string `long:"group"`
+
+	Positional struct {
+		LocalPath  string `positional-arg-name:"<local-path>" required:"1"`
+		RemotePath string `positional-arg-name:"<remote-path>" required:"1"`
+	} `positional-args:"yes"`
+}
+
+func init() {
+	AddCommand(&CmdInfo{
+		Name:        "push",
+		Summary:     cmdPushSummary,
+		Description: cmdPushDescription,
+		ArgsHelp: map[string]string{
+			"-p":      "Create parent directories for the file",
+			"-m":      "Override permissions (3-digit octal)",
+			"--uid":   "Use specified user ID",
+			"--user":  "Use specified username",
+			"--gid":   "Use specified group ID",
+			"--group": "Use specified group name",
+		},
+		New: func(opts *CmdOptions) flags.Commander {
+			return &cmdPush{client: opts.Client}
+		},
+	})
+}
+func (cmd *cmdPush) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	f, err := os.Open(cmd.Positional.LocalPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	st, err := f.Stat()
+	if err != nil {
+		return err
+	}
+
+	opts := client.PushOptions{
+		Source:      f,
+		Path:        cmd.Positional.RemotePath,
+		MakeDirs:    cmd.MakeDirs,
+		Permissions: st.Mode().Perm(),
+		UserID:      cmd.UserID,
+		User:        cmd.User,
+		GroupID:     cmd.GroupID,
+		Group:       cmd.Group,
+	}
+
+	if cmd.Permissions != "" {
```

Comment:
It fills a bit odd to always call stat and use `st.Mode().Perm()`, only to override it here. Let's structure this a bit differently, have a variable `var permissions os.FileMode` and only call stat in the else case where `cmd.Permissions == ""`:

```go
var permissions os.FileMode
if cmd.Permissions != "" {
    // ... parse it ...
    permissions = os.FileMode(p)
} else {
    // ... call stat ...
    permissions = st.Mode().Perm()
}

opts := client.PushOptions{
    // ...
}
```